features/nfc/doc/nfc_design.md

```diff
@@ -261,99 +261,190 @@ The `NFCRemoteTarget` class derives from `NFCTarget` and additionally from `NFCE
 
 ## NDEF API
 
-![ndef_diagram]
-
 The NDEF API is constructed with these requirements in mind:
 * Minimizing memory allocation/copies
 * NFC Forum compliance
 * Ease of use
+* Extensibility
 
-#### NDEF Message
+### Common objects
 
-A NDEF Message is made of multiple NDEF Records which is reflected by the API:
+We will provide multiple helpers to make it easy to create/parse common record types:
+* URI
+* Text
+* Smart Poster
+* MIME data
 
+For instance, the `URI`'s class API is as follows:
 ```cpp
-bool parse(const uint8_t* buffer, size_t sz)
-size_t count()
-NDEFRecord operator[](size_t n)
+uri_prefix_t uri_prefix() const
+void set_uri_prefix(uri_prefix_t prefix)
+
+bool get_uri(char* uri, size_t max_sz) const
+size_t uri_size() const
+void set_uri(const char* uri)
+
+bool get_full_uri(char* uri, size_t max_sz) const
+size_t full_uri_size() const
+void set_full_uri(const char* uri)
 ```
 
-The message can be mapped with a byte array and individual records are decoded/populated on the fly.
+**Note:** These types can be replaced by user defined ones if parsing and serialization logic is provided.
+
+### Parsing
 
-#### NDEF Message builder
+#### ndef::MessageParser
 
-We're using a builder pattern to encode an NDEF message over a byte array.
+![ndef_message_parser_diagram]
+
+Messages incoming from the peer are parsed by a `MessageParser` which produce
+`Record` instances to its client. The parsing operation is event-driven: a
+message parser client register a delegate inside the message parser. This delegate
+get notified whenever an interesting event happen during the parsing.
```

Comment:
--> The parsing operation is event-driven: a
 message parser client register**s** a delegate inside the message parser. This delegate
 get**s** notified whenever an interesting event happen**s** during the parsing.

Reply:
I wonder I never apply the correct form. It is not that difficult. 
Thanks.